### PR TITLE
feat(snql) Add condition validation

### PR DESCRIPTION
**Pull request by evanh**
In the legacy API, entities had extensions that would automatically add the
conditions that we required on Clickhouse. SnQL will not provide conditions
automatically, but we still need to verify that a query has the required
conditions.

Add required conditions to the entities, and a check in the SnQL parser to find
a top level condition for those columns. At the moment we expect the bare column
in the condition, which seems reasonable considering the current extensions.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -10,5 +10,7 @@ setup-git:
 test:
 	SNUBA_SETTINGS=test pytest -vv
 
+tests: test
```
Comment (evanh) [inline Makefile:13]: I added this because zsh constantly asks me if I really meant 'test' and I got tired of it.

File: snuba/datasets/entities/discover.py [hunk 2]
```diff
@@ -451,6 +452,15 @@ def selector_func(_query: Query) -> Tuple[str, List[str]]:
             ),
             join_relationships={},
             writable_storage=None,
+            required_conditions={
+                "project_id": [ConditionFunctions.EQ, ConditionFunctions.IN],
+                "timestamp": [
+                    ConditionFunctions.LT,
+                    ConditionFunctions.LTE,
+                    ConditionFunctions.GT,
+                    ConditionFunctions.GTE,
```
Comment (lynnagara) [inline snuba/datasets/entities/discover.py:461]: Why not EQ or IN?

**Review by lynnagara — COMMENTED** [review]
Just to confirm, does this mean those required conditions need to be provided as top level? That seems somewhat restrictive and potentially confusing to the end user. For example, it could be unexpected if `project_id IN [1,2,3]` is valid input but writing the condition as `project_id == 1 OR project_id = 2` doesn't work.

File: snuba/datasets/entities/discover.py (hunk 2 above)
Comment (evanh) [reply snuba/datasets/entities/discover.py:461]: Yeah I left those out because the extensions as they are now don't use those, but I'll add them in too.

**Comment by evanh** [comment]
> Just to confirm, does this mean those required conditions need to be provided as top level?

Yes it does. I know that is more restrictive than is technically necessary, but trying to validate all the edge cases is non-trivial. Allowing this also implies we should allow functions (e.g. `toString(project_id) = '1'`) but it gets too difficult to validate all the inputs.

TL;DR It is restrictive but I don't think that's a bad thing.

**Review by lynnagara — APPROVED** [review]

File: snuba/datasets/entities/sessions.py [hunk 3]
```diff
@@ -99,6 +100,18 @@ def __init__(self) -> None:
             abstract_column_set=read_schema.get_columns(),
             join_relationships={},
             writable_storage=writable_storage,
+            required_conditions={
+                "org_id": [ConditionFunctions.EQ],
```
Comment (lynnagara) [inline snuba/datasets/entities/sessions.py:104]: Would we want to check the type of the value we're comparing to? Like to ensure org_id is an int in this case, and not some other kind of expression

Comment (evanh) [reply snuba/datasets/entities/sessions.py:104]: I can. I left that off for two reasons: 1) we haven't started doing that type of checking yet 2) The query will fail if you pass a bad argument, and these are here to prevent valid queries that run over too much data.

Comment (lynnagara) [reply snuba/datasets/entities/sessions.py:104]: The query might still be valid as far as ClickHouse is concerned though, and cause ClickHouse to read a lot of data.
e.g. `org_id = key_id`

File: snuba/datasets/entity.py [hunk 4]
```diff
@@ -72,6 +74,9 @@ def get_all_join_relationships(self) -> Mapping[str, JoinRelationship]:
         """
         return self.__join_relationships
 
+    def get_required_conditions(self) -> Mapping[str, Sequence[str]]:
+        return self.__required_conditions
+
```
Comment (fpacifici) [inline snuba/datasets/entity.py:79]: Shouldn't this allow multiple condition requested for the same field? Like `timestamp` requiring a GTE and a LT condition to ensure we have an interval in the query.

File: snuba/datasets/entities/sessions.py (hunk 3 above)
Comment (fpacifici) [reply snuba/datasets/entities/sessions.py:104]: I'd agree with @lynnagara  on this. For conditions like project and organization, being overly strict is a good idea so I would also validate the right hand side of the condition to be of the right type. I am ok in doing it in a followup if it adds too much complexity in this one.

File: snuba/query/snql/parser.py [hunk 5]
```diff
@@ -871,6 +872,54 @@ def mangle_column_value(exp: Expression) -> Expression:
...
+        conditions = get_first_level_and_conditions(condition)
+        seen = {r: False for r in required}
+        for condition in conditions:
+            if isinstance(condition, FunctionCall) and condition.parameters:
+                param = condition.parameters[0]
+                if isinstance(param, Column) and seen.get(param.column_name) == False:
+                    if not table_name or param.table_name == table_name:
+                        allowed_ops = required[param.column_name]
+                        if condition.function_name in allowed_ops:
+                            seen[param.column_name] = True
```
Comment (fpacifici) [inline snuba/query/snql/parser.py:896]: Have you consider expressing the mandatory condition as matcher? That would give the flexibility of imposing conditions on the right hand side and you would not need custom code to validate each condition.

File: snuba/datasets/cdc/groupassignee_entity.py [hunk 6]
```diff
@@ -31,6 +31,8 @@ def __init__(self) -> None:
             abstract_column_set=schema.get_columns(),
             join_relationships={},
             writable_storage=storage,
+            required_filter_columns=None,
+            required_time_column=None,
```
Comment (lynnagara) [inline snuba/datasets/cdc/groupassignee_entity.py:35]: This is probably a distant concern but I guess it's at least theoretically possible for future conditions we may want to require on an entity to not fit neatly into one of these two categories. Maybe we want to allow GT/LT on something other than a time column, or somehow customise the project, org and time filters for an entity. I don't have any realistic use case right now since our data is essentially all time series by org or project, for which this works fine. Still I'm curious if this was a consideration when deciding on this format. Did you consider having the entity provide the matchers so they can be fully customised?

Comment (evanh) [reply snuba/datasets/cdc/groupassignee_entity.py:35]: Yes, that's why I moved the validation function into the Entity base class, and that function takes the entire query as an argument. Now if certain entities want custom validation, they can override the validation on the base class to do whatever custom work they want.

This actually came up as part of the sessions work, since we talked about having different required columns based on the granularity requested in the query. So the Sessions entity could override the validation, and have its own logic based on the granularity.

File: snuba/datasets/entity.py (hunk 4 above)
Comment (evanh) [reply snuba/datasets/entity.py]: Why are both necessary? Only one is needed to create a time window < 90 days.

File: snuba/datasets/entities/sessions.py (hunk 3 above)
Comment (evanh) [reply snuba/datasets/entities/sessions.py]: The matchers now do this.

File: snuba/datasets/entity.py (hunk 4 above)
Comment (fpacifici) [reply snuba/datasets/entity.py]: They are desirable so that clients have to be deliberate in picking a closed time range. In one way is to filter off spurious events that have a timestamp in the future (they exist, not kidding). And on the other hand is for when they want a short time range in the past (like 12 hours of one day 2 months ago). Having to provide the close interval pushes the client to think what is the shortest time range they need.
